sql/layers/layers.go

```diff
@@ -95,6 +95,55 @@ func UnsetAppliedFrom(db sql.Executor, lid types.LayerID) error {
 	return nil
 }
 
+// UpdateStateHash for the layer.
+func UpdateStateHash(db sql.Executor, lid types.LayerID, hash types.Hash32) error {
+	if _, err := db.Exec(`insert into layers (id, state_hash) values (?1, ?2) 
+	on conflict(id) do update set state_hash=?2;`,
+		func(stmt *sql.Statement) {
+			stmt.BindInt64(1, int64(lid.Value))
+			stmt.BindBytes(2, hash[:])
+		}, nil); err != nil {
+		return fmt.Errorf("set applied %s: %w", lid, err)
+	}
+	return nil
+}
+
+// GetLatestStateHash loads latest state hash.
+func GetLatestStateHash(db sql.Executor) (rst types.Hash32, err error) {
+	if rows, err := db.Exec("select state_hash from layers where state_hash is not null;",
```

Comment:
why isn't this querying the max layer id where state_hash is not null?

Reply:
layers are stored sorted by descending order, so we just go down and pick first where state_hash is not empty

Reply:
on second thoughts, i might be safer to explicitly choose the max in case of a schema change in the future?

Reply:
i think it is correct to rely on the order, moreover i dont know if adding a condition will result in the full scan or just pick the top layer. if it relies on the scan it will be a problem very fast. if someone accidentally changes schema there will be a test failure